### PR TITLE
Add Neighbors functions

### PR DESCRIPTION
**Pull request by richardartoul**
Reference implementations:

1) https://github.com/corsc/go-geohash/blob/master/geohash_int.go
2) https://github.com/gansidui/geohash/blob/master/geohash.go
3) https://github.com/yinqiwen/geohash-int/blob/master/geohash.c
4) https://github.com/sunng87/node-geohash/blob/master/main.js

Discussion points:

1) I thought this algorithm would be broken around the poles / equator / prime-meridian, but I think your encode function just handles that stuff gracefully? I'm a little confused why points near those areas work fine, but when I plotted the neighbors on a map it looked like it was doing the right thing.

2) Test cases all copied from here: https://github.com/sunng87/node-geohash/blob/master/tests/test.js (Any idea where to get better reference tests?)

3) I'm not confident this is the fastest implementation. I've seen other implementation that seem to be using different algorithms, but this is the only algorithm I could understand.

### CODE REVIEW

**Comment by richardartoul** [comment]
Ok I auto-generated some tests. Copy paste of the comment I left in the auto-genned test file:

# Test cases generated with ngeohash - Note that I had to remove some of the test cases because they varied slightly (like by really really tiny amounts) from the results generated by this
library. I did some digging and the results were always so close (even if you plot on a Map you can't really tell the difference) that I couldn't comeup with an explanation other than the fact that Javascript uses 64-bit floating point numbers for everything while Golang has support for true integers. If you look through the ngeohash code you'll see that they do many operations on floats that could result in tiny losses of precision.

**Comment by mmcloughlin** [comment]
Yes, I have observed that for the other geohash tests. I believe it is a result of both floating point precision, and a different method of mapping lat/lng values to integers. In this library we just use integer rounding (https://github.com/mmcloughlin/geohash/blob/master/geohash.go#L137-L141), but the description of the algorithm describes a binary-search-like technique (https://en.wikipedia.org/wiki/Geohash). I have no idea why they choose to describe it this way, it seems far more complicated to me. But I can see how even though both methods are theoretically equivalent, the combination with floating point precision issues could lead to subtle differences in low bits.

* That said, it does subtly bother me that the file of test cases is not reproducible. What do you think? Is there a way we can avoid this problem, perhaps by limiting ourselves to lower precision?

* Also note that the other generated test cases writes its own source code into the output https://github.com/mmcloughlin/geohash/blob/master/testcases_test.go#L15. This is related to making the file generation reproducible.

File: neighbors_testcases_test.go [hunk 1]
```diff
@@ -0,0 +1,514 @@
...
+
+//	# Test cases generated with ngeohash - Note that I had to remove some of the
+//  test cases because they varied slightly from the results generated by this
+//  library. I did some digging and the results were always so close (even if
+//  you plot on a Map you can't really tell the difference) that I couldn't come
+//  up with an explanation other than the fact that Javascript uses 64-bit
+//  floating point numbers for everything while Golang has support for true
+//  uint64's. If you look through the ngeohash code you'll see that they do
+//  many operations on floats that could result in tiny losses of precision.
+// Note that I had to remove
```
Comment (mmcloughlin) [inline neighbors_testcases_test.go]: Repeat?

File: neighbors_testcases_test.go [hunk 2]
```diff
@@ -0,0 +1,514 @@
...
+//   let lat = (-90 + 180 * Math.random())
+//   // -180 --> 180
+//   let lng = (-180 + 360 * Math.random())
+//
+//   // 1 --> 9
+//   let strPrecision = Math.floor(Math.random() * 9) + 1
+//   // 2 --> 52. NGeohash has a max precision of 52 (because Javascript)
+//   let intBitdepth = Math.floor(Math.random() * 52) + 2
+//   // ngeohash library specifies that bitdepth must be even
+//   if (intBitdepth %2 != 0) {
```
Comment (mmcloughlin) [inline neighbors_testcases_test.go]:
```javascript
let intBitdepth = 2 * Math.floor(Math.random() * 26) + 2
```

File: neighbors_testcases_test.go [hunk 3]
```diff
@@ -0,0 +1,514 @@
+package geohash
+
+//	# Test cases generated with ngeohash - Note that I had to remove some of the
```
Comment (mmcloughlin) [inline neighbors_testcases_test.go]: Indent? As mentioned in other comment, perhaps have your node code write its own source code into the output file?

File: geohash_test.go [hunk 4]
```diff
@@ -131,3 +131,60 @@ func TestLeadingZero(t *testing.T) {
...
 	}
 }
+
+func TestNeighbors(t *testing.T) {
+	for _, c := range neighborsTestCases {
+		neighbors := Neighbors(c.hashStr)
+		for i, neighbor := range neighbors {
+			expected := c.hashStrNeighbors[i]
+			if neighbor != expected {
+				t.Errorf("Actual: %v \n expected: %v\n", neighbors, c.hashIntNeighbors)
```
Comment (mmcloughlin) [inline geohash_test.go]: Very nitpicky but all the other failure messages are lowercase.

File: neighbors_testcases_test.go [hunk 5]
```diff
@@ -0,0 +1,514 @@
+package geohash
```
Comment (mmcloughlin) [inline neighbors_testcases_test.go]: Does this output file pass `go fmt`?

**Comment by mmcloughlin** [comment]
We also need to regenerate `README.md` from `README.tmpl`. This was done with https://github.com/robertkrimen/godocdown.

File: geohash.go [hunk 6]
```diff
@@ -134,6 +134,67 @@ func DecodeInt(hash uint64) (lat, lng float64) {
...
+		// SW,
+		EncodeWithPrecision(lat-latDelta, lng-lngDelta, precision),
+		// W,
+		EncodeWithPrecision(lat, lng-lngDelta, precision),
+		// NW
+		EncodeWithPrecision(lat+latDelta, lng-lngDelta, precision),
+	}
+}
+
+// NeighborsInt returns a slice of uint64's that correspond to the provided hash's
```
Comment (mmcloughlin) [inline geohash.go]: `uint64's` -> `uint64s` ?

File: geohash.go [hunk 7]
```diff
@@ -134,6 +134,67 @@ func DecodeInt(hash uint64) (lat, lng float64) {
...
+// neighbors at 64-bit precision.
+func NeighborsInt(hash uint64) []uint64 {
+	return NeighborsIntWithPrecision(hash, 64)
+}
+
+// NeighborsIntWithPrecision returns a slice of uint64's that correspond to the
```
Comment (mmcloughlin) [inline geohash.go]: `uint64's` -> `uint64s` ?

File: neighbors_testcases_test.go (hunk 2 above)
Comment (richardartoul) [inline neighbors_testcases_test.go]: Oh nice thats clever.

File: neighbors_testcases_test.go (hunk 5 above)
Comment (richardartoul) [inline neighbors_testcases_test.go]: I'll verify.

File: neighbors_testcases_test.go (hunk 3 above)
Comment (richardartoul) [inline neighbors_testcases_test.go]: Will fix.

File: neighbors_testcases_test.go (hunk 1 above)
Comment (richardartoul) [inline neighbors_testcases_test.go]: Will fix.

File: geohash_test.go (hunk 4 above)
Comment (richardartoul) [inline geohash_test.go]: Will fix

File: neighbors_testcases_test.go [hunk 8]
```diff
@@ -0,0 +1,1084 @@
...
+//	     return line === "" ? `//` : `//\t ${line}`;
+//	   }).join("\n")
+//	 }
+//
+//	 // Native JS Math.Rand() does not support seeds.
+//	 // Solution from: http://stackoverflow.com/questions/521295/javascript-random-seeds
+//	 function getSeededRand(seed) {
+//	   return function randWithSeed() {
+//	 		var x;
+//	     do {
```
Comment (richardartoul) [inline neighbors_testcases_test.go]: Looks weird in the preview but if you pull down the code it looks normal

**Comment by richardartoul** [comment]
@mmcloughlin Addressed all your feedback (thanks for the thoughtful comments.) Let me know if there are any other changes you'd like, but I think its starting to look pretty good.

**Pull request merged**